### PR TITLE
Push CBC_BREAKPOINT to the opcode list

### PR DESCRIPTION
**Pull request by polaroi8d**
Add breakpoints:

- active - `(CBC_BREAKPOINT_ENABLED)`  when the additional breakpoint is active, the engine stop running on the breakpoint

- inactive - `(CBC_BREAKPOINT_DISABLED)` and  there is nothing to do when the engine step in the breakpoint

JerryScript-DCO-1.0-Signed-off-by: Levente Orban orbanl@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/parser/js/js-parser.c [hunk 1]
```diff
@@ -1850,6 +1850,10 @@ parser_parse_source (const uint8_t *source_p, /**< valid UTF-8 source code */
   context.context_stack_depth = 0;
 #endif /* !JERRY_NDEBUG */
 
+#ifdef DEBUGGER
+  context.statement_line = 0;
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser.c]: When this context.statement_line is updated? I only saw one setting here.

File: jerry-core/parser/js/js-parser-statm.c [hunk 2]
```diff
@@ -1641,6 +1641,20 @@ parser_parse_statements (parser_context_t *context_p) /**< context */
   while (context_p->token.type != LEXER_EOS
          || context_p->stack_top_uint8 != PARSER_STATEMENT_START)
   {
+#ifdef DEBUGGER
+    if (context_p->line != context_p->statement_line)
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-statm.c]: There is no check whether debugger is enabled during init.

Comment (polaroi8d) [reply jerry-core/parser/js/js-parser-statm.c]: @zherczeg you think about this one `if (JERRY_CONTEXT (jerry_init_flags) & JERRY_INIT_DEBUGGER)` ?

File: jerry-core/parser/js/js-parser-statm.c [hunk 3]
```diff
@@ -1889,6 +1889,15 @@ parser_parse_statements (parser_context_t *context_p) /**< context */
         break;
       }
     }
+    /* Add breakpoints, after every single line. This breakpoints has an
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-statm.c]: Breakpoints should be added before every line :) When the line starts.

File: jerry-core/parser/js/js-parser-statm.c [hunk 4]
```diff
@@ -1889,6 +1889,15 @@ parser_parse_statements (parser_context_t *context_p) /**< context */
...
+    *  status property which can hold two type. One of them the active,
+    *  when the additional breakpoint is active, the engine stop running on
+    *  the breakpoint. The other option is when the breakpoint is inactive and
+    *  there is nothing to do when the engine step in the breakpoint.
+    */
+#ifdef JERRY_DEBUGGER
+    parser_emit_cbc (context_p, CBC_BREAKPOINT);
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-statm.c]: First, we need to check whether debugging is enabled, and we should insert NOP-s, which can be turned to debugger at any point by the debugger.

Comment (polaroi8d) [reply jerry-core/parser/js/js-parser-statm.c]: You think to add a `parser_emit_cbc(context_p, CBC_EXT_NOP);` ?

File: jerry-core/parser/js/js-parser-statm.c (hunk 3 above)
Comment (polaroi8d) [reply jerry-core/parser/js/js-parser-statm.c]: Okey.

File: jerry-core/parser/js/byte-code.h [hunk 5]
```diff
@@ -316,8 +316,10 @@
               VM_OC_RET) \
   CBC_OPCODE (CBC_RETURN_WITH_LITERAL, CBC_HAS_LITERAL_ARG, 0, \
               VM_OC_RET | VM_OC_GET_LITERAL) \
-  CBC_OPCODE (CBC_BREAKPOINT, CBC_NO_FLAG, 0, \
-              VM_OC_BREAKPOINT) \
+  CBC_OPCODE (CBC_BREAKPOINT_ENABLE, CBC_NO_FLAG, 0, \
+              VM_OC_BREAKPOINT_ENABLE) \
+  CBC_OPCODE (CBC_BREAKPOINT_DESABLE, CBC_NO_FLAG, 0, \
```
Comment (zherczeg) [inline jerry-core/parser/js/byte-code.h]: CBC_BREAKPOINT_ENABLED CBC_BREAKPOINT_DISABLED

File: jerry-core/parser/js/js-parser-statm.c [hunk 6]
```diff
@@ -1656,6 +1656,21 @@ parser_parse_statements (parser_context_t *context_p) /**< context */
     JERRY_ASSERT (context_p->stack_depth == context_p->context_stack_depth);
 #endif /* !JERRY_NDEBUG */
 
+    /* Add breakpoints, before every single line. We have two tpye of breakpoints
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-statm.c]: type

File: jerry-core/parser/js/js-parser-statm.c [hunk 7]
```diff
@@ -1656,6 +1656,21 @@ parser_parse_statements (parser_context_t *context_p) /**< context */
...
+    *  one of them the active (enable) others the inactive (desable). In basic
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-statm.c]: enabled, disabled

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: jerry-core/parser/js/js-parser-statm.c (hunk 7 above)
Comment (polaroi8d) [reply jerry-core/parser/js/js-parser-statm.c]: Thanks for the review, updated the PR.

File: jerry-core/parser/js/js-parser-statm.c [hunk 8]
```diff
@@ -1656,6 +1656,21 @@ parser_parse_statements (parser_context_t *context_p) /**< context */
...
+    *  state the breakpoints is desabled. When the additional breakpoint is active,
+    *  the engine stop running on the breakpoint. The other option is when the
+    *  breakpoint is inactive and there is nothing to do when the engine step
+    *  in the breakpoint.
+    */
+#ifdef JERRY_DEBUGGER
+    if (JERRY_CONTEXT (jerry_init_flags) & JERRY_INIT_DEBUGGER)
+    {
+      parser_emit_cbc (context_p, CBC_BREAKPOINT_DISABLED);
+      parser_flush_cbc(context_p);
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-statm.c]: It seems you add a breakpoint before each statement. You should only do it before each line according to the comment above.

File: jerry-core/parser/js/js-parser-internal.h [hunk 9]
```diff
@@ -271,6 +271,10 @@ typedef struct
   int is_show_opcodes;                        /**< show opcodes */
   uint32_t total_byte_code_size;              /**< total byte code size */
 #endif /* PARSER_DUMP_BYTE_CODE */
+
+#ifdef JERRY_DEBUGGER
+  parser_line_counter_t last_bp_line;         /**< last line where breakpoint was inserted */
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-internal.h]: last_breakpoint_line

File: jerry-core/parser/js/js-parser-statm.c [hunk 10]
```diff
@@ -1656,6 +1682,29 @@ parser_parse_statements (parser_context_t *context_p) /**< context */
...
+    *  inactive and there is nothing to do when the engine step in the breakpoint. */
+#ifdef JERRY_DEBUGGER
+    if (JERRY_CONTEXT (jerry_init_flags) & JERRY_INIT_DEBUGGER)
+    {
+      if (context_p->line != context_p->last_bp_line
+          && context_p->token.type != LEXER_SEMICOLON
+          && context_p->token.type != LEXER_LEFT_BRACE
+          && context_p->token.type != LEXER_RIGHT_BRACE
+          && context_p->token.type != LEXER_KEYW_VAR
+          && context_p->token.type != LEXER_KEYW_FUNCTION)
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-statm.c]: Add case and default keywords here.

**Review by zherczeg — CHANGES_REQUESTED** [review]

**Comment by zherczeg** [comment]
LGTM

**Pull request merged**